core/jvm/src/main/java/zio/internal/Mailbox.java

```diff
@@ -0,0 +1,146 @@
+ *           Non-intrusive MPSC node-based queue</a> by D. Vyukov.
+ */
+final class Mailbox<A> extends AtomicReference<Mailbox.Node<A>> {
+
+	private transient Node<A> read;
+
+	Mailbox() {
+		super(new Node<A>(null));
+		read = getPlain();
+	}
+
+	/**
+	 * Adds the specified element to the queue.
+	 * 
+	 * @throws NullPointerException if element is {@code null}
+	 */
+	public void add(A data) throws NullPointerException {
+		if (null == data)
+			throw new NullPointerException();
+		Node<A> next = new Node<A>(data);
+		Node<A> prev = getAndSet(next);
+		prev.lazySet(next);
+	}
+
+	/**
+	 * Returns {@code true} if the queue has no elements. Otherwise, returns
+	 * {@code false}.
+	 * 
+	 * @apiNote This method SHOULD be invoked by the single consumer (thread).
+	 * 
+	 * @apiNote This method can return a false negative at some point if elements
+	 *          were added using `prepend`. Check the value returned by `poll` in
+	 *          such cases.
+	 */
+	public boolean isEmpty() {
+		return null == read.get();
+	}
+
+	/**
+	 * Returns {@code true} if the queue has elements. Otherwise, returns
+	 * {@code false}.
+	 * 
+	 * @apiNote This method SHOULD be invoked by the single consumer (thread).
+	 *
+	 * @apiNote This method can return a false positive at some point if elements
+	 *          were added using `prepend`. Check the value returned by `poll` in
+	 *          such cases.
+	 */
+	public boolean nonEmpty() {
+		return null != read.get();
+	}
+
+	/**
+	 * Adds the specified element to the beginning of the queue; the element will be
+	 * read before any other element in the queue.
+	 *
+	 * @apiNote This method MUST be invoked by the single consumer (thread).
+	 * 
+	 * @throws NullPointerException if element is {@code null}
+	 */
+	public void prepend(A data) throws NullPointerException {
+		if (null == data)
+			throw new NullPointerException();
+		// optimization: avoid atomic operation by retaining (phantom) read node
+		read = new Node<A>(null, new Node<A>(data, read));
+	}
+
+	/**
+	 * A version of `prepend` that adds 2 elements to the beginning of the queue.
+	 *
+	 * Calling this method is equivalent to running
+	 * 
+	 * <pre>
+	 * mailbox.prepend(data2);
+	 * mailbox.prepend(data1);
+	 * </pre>
+	 *
+	 * @apiNote This method MUST be invoked by the single consumer (thread).
+	 * 
+	 * @throws NullPointerException if either element is {@code null}
+	 */
+	public void prepend2(A data1, A data2) throws NullPointerException {
+		if (null == data1 || null == data2)
+			throw new NullPointerException();
+		// optimization: avoid atomic operation by retaining (phantom) read node
+		read = new Node<A>(null, new Node<A>(data1, new Node<A>(data2, read)));
+	}
+
+	/**
+	 * Removes and returns the oldest element in the queue if the queue has
+	 * elements. Otherwise, returns {@code null}.
+	 * 
+	 * @apiNote This method MUST be invoked by the single consumer (thread).
+	 */
+	public A poll() {
+		Node<A> next;
+		A data;
+		while (true) {
```

Comment:
notice this loops is limited to two cycles, 99% of the time it'd only have 1.
make sure the loop doesn't introduce unnecessary cost (I used a simple 'recursive' call to poll, knowing it will never accumulate more than 1 stack frame)

Reply:
Removed loop